cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -64,12 +65,17 @@ class hostdevice_vector {
 
   T &operator[](size_t i) const { return h_data[i]; }
   T *host_ptr(size_t offset = 0) const { return h_data + offset; }
-  T *device_ptr(size_t offset = 0) const { return d_data + offset; }
+  T *device_ptr(size_t offset = 0) { 
+      return reinterpret_cast<T*>(d_data.data()) + offset; 
+    }
+  T const* device_ptr(size_t offset = 0) const { 
+      return reinterpret_cast<T const*>(d_data.data()) + offset; 
+    }
 
  private:
   cudaStream_t stream = 0;
   size_t max_elements = 0;
   size_t num_elements = 0;
   T *h_data = nullptr;
-  T *d_data = nullptr;
+  rmm::device_buffer d_data;
```

Comment:
If the data is typed (previously was a T*), then should this be a device_vector rather than a device_buffer?

Reply:
I used device_vector at first, but then realized that I can't pass the stream. I used device_buffer wherever stream parameter was used for allocation, to avoid changing the synchronization.

Reply:
Doesn't device_vector also implies zero-initialization ? (we definitely don't want that)

Reply:
It presumably does, AFAIK they imitate the std::vector design, but I don't see this info in their docs.

Reply:
OK, those are good reasons. We do have an open issue to create an RMM device_vector that solves those issues. But currently device_buffer is the right choice.